src/Integration.Tests/CoroutineTests.cs

```diff
@@ -29,6 +29,65 @@ public async Task MultipleCoroutineCalls()
         Assert.All(r, x => Assert.Equal(5, x));
     }
 
+    [Fact]
+    public async Task SequentialCoroutinesWithCompletedCancellation()
+    {
+        var mod = Env.TestCoroutines();
+
+        using CancellationTokenSource cts = new();
+
+        async Task RunAsync()
+        {
+            _ = await mod.TestCoroutine(seconds: 0, cancellationToken: cts.Token)
+                         .WaitAsync(TimeSpan.FromSeconds(10), TestContext.Current.CancellationToken);
+        }
+
+        // First call should complete successfully.
+
+        await RunAsync();
+
+        cts.Cancel();
+
+        // Second call should be cancelled immediately (implicitly tests that
+        // the event loop is still functional).
+
+        await Assert.ThrowsAsync<TaskCanceledException>(() => RunAsync());
+    }
+
+    [Fact]
+    public async Task SequentialCoroutinesWithErrorCancellation()
+    {
+        var mod = Env.TestCoroutines();
+
+        using CancellationTokenSource cts = new();
+
+        async Task RunAsync()
+        {
+            _ = await mod.TestCoroutineRaisesException(cancellationToken: cts.Token)
+                         .WaitAsync(TimeSpan.FromSeconds(10), TestContext.Current.CancellationToken);
+        }
+
+        // First call should throw an error.
+
+        await Assert.ThrowsAsync<PythonInvocationException>(() => RunAsync());
+
+        cts.Cancel();
+
+        // Second call should be cancelled immediately (implicitly tests that
+        // the event loop is still functional).
+
+        await Assert.ThrowsAsync<TaskCanceledException>(() => RunAsync());
+    }
+
+    [Fact]
+    public async Task CoroutineThatSelfCancels()
+    {
+        var mod = Env.TestCoroutines();
+        var task = mod.TestCoroutineSelfCanceling(TestContext.Current.CancellationToken);
+        await Assert.ThrowsAsync<TaskCanceledException>(() => task);
```

Comment:
Two things:

- Why not use a simple local function to be the `Act` here and elsewhere for less task/lambda ceremony?
- I would make the strong assertion that the cancellation wasn't due to some `CancellationToken` so chances are good that the source was something else (Python code in this case).

```suggestion
        Task Act() => mod.TestCoroutineSelfCanceling(TestContext.Current.CancellationToken);
        var ex = await Assert.ThrowsAsync<TaskCanceledException>(Act);
        Assert.Equal(CancellationToken.None, ex.CancellationToken);
```